File: 05-browser-extension/shared/fingerprint.js
// Shared ID-derivation logic. MUST stay in sync with
// 06-inventory-crawler/fingerprint.py and extract.js so that observed
// automation events line up with the crawled inventory.

const SLUG_RE = /[^a-z0-9]+/g;

export function slug(text) {
  return (text || "").trim().toLowerCase().replace(SLUG_RE, "-").replace(/^-+|-+$/g, "");
}

// Small, dependency-free hash to match Python's short_hash usage as a last resort.
export function shortHash(value, length = 8) {
  let h = 0;
  for (let i = 0; i < (value || "").length; i++) {
    h = (Math.imul(31, h) + value.charCodeAt(i)) | 0;
  }
  return (h >>> 0).toString(16).padStart(length, "0").slice(0, length);
}

const DEFAULT_DYNAMIC = [/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-/, /^\d+$/, /^[0-9a-f]{24}$/];

export function normalizePath(pathname, dynamicPatterns = DEFAULT_DYNAMIC) {
  let path = (pathname || "/").split("?")[0].split("#")[0];
  if (!path.startsWith("/")) path = "/" + path;
  path = path.replace(/\/+/g, "/").replace(/\/+$/, "") || "/";
  return (
    path
      .split("/")
      .map(part => {
        if (!part) return part;
        return dynamicPatterns.some(rx => rx.test(part)) ? "{id}" : part;
      })
      .join("/") || "/"
  );
}

export function routeIdFromPath(normalizedPath) {
  if (normalizedPath === "" || normalizedPath === "/") return "route-root";
  const body = slug(normalizedPath.replace(/\{id\}/g, "id"));
  return `route-${body || "root"}`;
}

export function elementKey(el) {
  const testid = el.getAttribute("data-testid");
  const coverageId = el.getAttribute("data-coverage-id") || el.getAttribute("data-coverage-action");
  if (testid) return `testid:${testid}`;
  if (coverageId) return `coverageId:${coverageId}`;
  if (el.id) return `id:${el.id}`;
  const name = el.getAttribute("name");
  if (name) return `name:${name}`;
  const aria = el.getAttribute("aria-label");
  if (aria) return `aria:${slug(aria)}`;
  const role = el.getAttribute("role") || el.tagName.toLowerCase();
  const text = slug((el.textContent || "").replace(/\s+/g, " ").trim()).slice(0, 40);
  if (text) return `${role}:${text}`;
  return `path:${shortHash(cssPath(el))}`;
}

export function cssPath(node) {
  const parts = [];
  let el = node;
  while (el && el.nodeType === 1 && parts.length < 6) {
    let seg = el.tagName.toLowerCase();
    if (el.id) {
      seg += `#${el.id}`;
      parts.unshift(seg);
      break;
    }
    const parent = el.parentElement;
    if (parent) {
      const sameTag = [...parent.children].filter(c => c.tagName === el.tagName);
      if (sameTag.length > 1) seg += `:nth-of-type(${sameTag.indexOf(el) + 1})`;
    }
    parts.unshift(seg);
    el = el.parentElement;
  }
  return parts.join(" > ");
}

function containerKey(el) {
  const landmark = el.closest(
    "form, nav, section, aside, header, main, [role='region'], [data-coverage-id], [data-testid]"
  );
  if (!landmark) return "page";
  return (
    landmark.getAttribute("data-testid") ||
    landmark.getAttribute("data-coverage-id") ||
    landmark.getAttribute("aria-label") ||
    landmark.id ||
    (landmark.className && String(landmark.className).split(" ")[0]) ||
    landmark.tagName.toLowerCase()
  );
}

export function componentIdFor(normalizedPath, el) {
  const rid = routeIdFromPath(normalizedPath).slice(6);
  const ckey = slug(containerKey(el)) || shortHash(containerKey(el));
  return `component-${rid}-${ckey}`;
}

export function actionIdFor(normalizedPath, el) {
  const rid = routeIdFromPath(normalizedPath).slice(6);
  return `action-${rid}-${slug(elementKey(el))}`;
}

File: 06-inventory-crawler/fingerprint.py
"""Deterministic fingerprinting for routes, components, and actions.

The crawler and the browser extension MUST derive identical IDs so that
observed automation events line up with the discovered inventory. Keep the
rules in this module in sync with `shared/fingerprint.js` (used by the
extension). Any change to normalization here must be mirrored there.
"""
from __future__ import annotations

import hashlib
import re
from typing import Iterable

_SLUG_RE = re.compile(r"[^a-z0-9]+")


def slug(text: str) -> str:
    """Lowercase, hyphenated, trimmed slug."""
    text = (text or "").strip().lower()
    text = _SLUG_RE.sub("-", text)
    return text.strip("-")


def short_hash(value: str, length: int = 8) -> str:
    return hashlib.sha1(value.encode("utf-8")).hexdigest()[:length]


def normalize_path(pathname: str, dynamic_patterns: Iterable[str]) -> str:
    """Collapse dynamic URL segments so /orders/123 == /orders/{id}."""
    compiled = [re.compile(p) for p in dynamic_patterns]
    pathname = (pathname or "/").split("?")[0].split("#")[0]
    if not pathname.startswith("/"):
        pathname = "/" + pathname
    pathname = re.sub(r"/+", "/", pathname).rstrip("/") or "/"
    parts = pathname.split("/")
    out = []
    for part in parts:
        if not part:
            out.append(part)
            continue
        if any(rx.search(part) for rx in compiled):
            out.append("{id}")
        else:
            out.append(part)
    return "/".join(out) or "/"


def route_id(normalized_path: str) -> str:
    if normalized_path in ("", "/"):
        return "route-root"
    body = slug(normalized_path.replace("{id}", "id"))
    return f"route-{body or 'root'}"


def element_key(el: dict) -> str:
    """Stable identity for an element using a fallback chain.

    el is a dict of attributes captured in the browser:
    testid, coverageId, id, name, ariaLabel, role, tag, text, cssPath.
    """
    for attr in ("testid", "coverageId"):
        if el.get(attr):
            return f"{attr}:{el[attr]}"
    if el.get("id"):
        return f"id:{el['id']}"
    if el.get("name"):
        return f"name:{el['name']}"
    if el.get("ariaLabel"):
        return f"aria:{slug(el['ariaLabel'])}"
    role = el.get("role") or el.get("tag") or "el"
    text = slug(el.get("text", ""))[:40]
    if text:
        return f"{role}:{text}"
    # Last resort: hash the CSS path so it is at least stable per render.
    return f"path:{short_hash(el.get('cssPath', ''))}"


def component_id(route_norm_path: str, container_key: str) -> str:
    return f"component-{route_id(route_norm_path)[6:]}-{slug(container_key) or short_hash(container_key)}"


def action_id(route_norm_path: str, el: dict) -> str:
    return f"action-{route_id(route_norm_path)[6:]}-{slug(element_key(el))}"


def action_event(tag: str, el_type: str | None) -> str:
    tag = (tag or "").lower()
    el_type = (el_type or "").lower()
    if tag == "form":
        return "submit"
    if tag in ("select", "textarea"):
        return "change"
    if tag == "input":
        if el_type in ("submit", "button", "checkbox", "radio"):
            return "click"
        return "change"
    return "click"
